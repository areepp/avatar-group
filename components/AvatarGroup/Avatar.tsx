import Image from 'next/image'
import classNames from 'classnames'

interface IPerson {
  id: number
  name: string
  initial: string
  imageSrc: string
}

interface Props {
  size: 'xs' | 'sm' | 'md' | 'lg'
  data: IPerson | number
}

const Avatar = ({ size, data }: Props) => {
  let children
  if (typeof data === 'object') {
    // render avatar image or the initial name if there is no image present
    children =
      data.imageSrc !== '' ? (
        <Image src={data.imageSrc} fill alt={data.name} />
      ) : (
        data.initial
      )
  } else {
    // render the number of remaining people
    children = `+${data}`
  }

  return (
    <div
      className={classNames(
        size === 'xs' && 'w-avatar-xs h-avatar-xs',
        size === 'sm' && 'w-avatar-sm h-avatar-sm',
        size === 'md' && 'w-avatar-md h-avatar-md',
        size === 'lg' && 'w-avatar-lg h-avatar-lg',
        'relative flex items-center justify-center bg-neutral-grey text-neutral-black rounded-full overflow-hidden border border-white',
      )}
    >
      {children}
    </div>
  )
}

export default Avatar
