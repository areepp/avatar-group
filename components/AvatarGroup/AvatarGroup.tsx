import Avatar from './Avatar'
import PEOPLE from '@/data/people.json'

interface Props {
  size: 'xs' | 'sm' | 'md' | 'lg'
  maxLength: number
}

const AvatarGroup = ({ size, maxLength }: Props) => {
  const showedPeople = PEOPLE.slice(0, maxLength)
  const remainingPeople = PEOPLE.length - showedPeople.length

  if (!showedPeople) return null

  return (
    <div className="flex -space-x-3">
      {showedPeople.map((person) => (
        <Avatar size={size} data={person} key={person.id} />
      ))}
      {remainingPeople > 0 && <Avatar size={size} data={remainingPeople} />}
    </div>
  )
}

export default AvatarGroup
