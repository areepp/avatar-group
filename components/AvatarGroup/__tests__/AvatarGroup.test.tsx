import { render, screen } from '@testing-library/react'
import AvatarGroup from '../AvatarGroup'
import '@testing-library/jest-dom'

// MOCK DATA
// [
//   {
//     "id": 1,
//     "name": "Tomi Budi",
//     "initial": "TM",
//     "imageSrc": "/profile-picture.jpg"
//   },
//   {
//     "id": 2,
//     "name": "Cahya",
//     "initial": "CH",
//     "imageSrc": ""
//   },
//   {
//     "id": 3,
//     "name": "Sri Mulyani",
//     "initial": "SM",
//     "imageSrc": "/profile-picture.jpg"
//   }
// ]

describe('AvatarGroup', () => {
  it('renders properly', () => {
    const { container } = render(<AvatarGroup maxLength={2} size="xs" />)

    expect(container).toMatchSnapshot()
  })

  it('shows avatar image or initial name', () => {
    render(<AvatarGroup maxLength={3} size="xs" />)

    expect(screen.getAllByRole('img')).toHaveLength(2)
    expect(screen.getByText('CH')).toBeInTheDocument()
  })

  it('shows +1 when the maxLength is 2', () => {
    render(<AvatarGroup maxLength={2} size="xs" />)

    expect(screen.getByText('+1')).toBeInTheDocument()
  })
})
