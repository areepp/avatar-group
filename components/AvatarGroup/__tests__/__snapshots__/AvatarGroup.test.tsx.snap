// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup renders correctly 1`] = `
<div>
  <div
    class="flex -space-x-3"
  >
    <div
      class="w-avatar-xs h-avatar-xs relative flex items-center justify-center bg-neutral-grey text-neutral-black rounded-full overflow-hidden border border-white"
    >
      <img
        alt="Tomi Budi"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Fprofile-picture.jpg&w=3840&q=75"
        srcset="/_next/image?url=%2Fprofile-picture.jpg&w=640&q=75 640w, /_next/image?url=%2Fprofile-picture.jpg&w=750&q=75 750w, /_next/image?url=%2Fprofile-picture.jpg&w=828&q=75 828w, /_next/image?url=%2Fprofile-picture.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fprofile-picture.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fprofile-picture.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fprofile-picture.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fprofile-picture.jpg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="w-avatar-xs h-avatar-xs relative flex items-center justify-center bg-neutral-grey text-neutral-black rounded-full overflow-hidden border border-white"
    >
      CH
    </div>
    <div
      class="w-avatar-xs h-avatar-xs relative flex items-center justify-center bg-neutral-grey text-neutral-black rounded-full overflow-hidden border border-white"
    >
      +1
    </div>
  </div>
</div>
`;

exports[`AvatarGroup renders properly 1`] = `
<div>
  <div
    class="flex -space-x-3"
  >
    <div
      class="w-avatar-xs h-avatar-xs relative flex items-center justify-center bg-neutral-grey text-neutral-black rounded-full overflow-hidden border border-white"
    >
      <img
        alt="Tomi Budi"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Fprofile-picture.jpg&w=3840&q=75"
        srcset="/_next/image?url=%2Fprofile-picture.jpg&w=640&q=75 640w, /_next/image?url=%2Fprofile-picture.jpg&w=750&q=75 750w, /_next/image?url=%2Fprofile-picture.jpg&w=828&q=75 828w, /_next/image?url=%2Fprofile-picture.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fprofile-picture.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fprofile-picture.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fprofile-picture.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fprofile-picture.jpg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="w-avatar-xs h-avatar-xs relative flex items-center justify-center bg-neutral-grey text-neutral-black rounded-full overflow-hidden border border-white"
    >
      CH
    </div>
    <div
      class="w-avatar-xs h-avatar-xs relative flex items-center justify-center bg-neutral-grey text-neutral-black rounded-full overflow-hidden border border-white"
    >
      +1
    </div>
  </div>
</div>
`;
